store/copr/coprocessor.go

```diff
@@ -95,8 +95,27 @@ func (c *CopClient) Send(ctx context.Context, req *kv.Request, variables interfa
 	ctx = context.WithValue(ctx, tikv.TxnStartKey(), req.StartTs)
 	ctx = context.WithValue(ctx, util.RequestSourceKey, req.RequestSource)
 	bo := backoff.NewBackofferWithVars(ctx, copBuildTaskMaxBackoff, vars)
-	ranges := NewKeyRanges(req.KeyRanges)
-	tasks, err := buildCopTasks(bo, c.store.GetRegionCache(), ranges, req, eventCb)
+	var (
+		tasks []*copTask
+		err   error
+	)
+	if len(req.KeyRanges) > 0 {
+		ranges := NewKeyRanges(req.KeyRanges)
+		tasks, err = buildCopTasks(bo, c.store.GetRegionCache(), ranges, req, eventCb)
+	} else {
+		// Here we build the task by partition, not directly by region.
+		// This is because it's possible that TiDB merge multiple small partition into one region which break some assumption.
+		// Keep it split by partition would be more safe.
+		for _, kvRanges := range req.KeyRangesWithPartition {
+			ranges := NewKeyRanges(kvRanges)
+			var tasksInPartition []*copTask
+			tasksInPartition, err = buildCopTasks(bo, c.store.GetRegionCache(), ranges, req, eventCb)
+			if err != nil {
+				break
+			}
+			tasks = append(tasks, tasksInPartition...)
```

Comment:
buildCopTasks handle the `req.Desc` property (reverse internally). When all key ranges are as a whole, that works.

If you split tasks for each partition, you make take extra care for the `req.Desc` case.

Reply:
There's an extra topn upon the reader. So the correctness is ensured. I'll add some more check.